lib/Expression_gen.ml

```diff
@@ -0,0 +1,262 @@
+open Core_kernel
+open Mir
+
+let ends_with suffix s =
+  String.is_suffix ~suffix:suffix s
+
+let starts_with prefix s =
+  String.is_prefix ~prefix:prefix s
+
+let functions_requiring_namespace = Set.Poly.of_list
+  [ "e"; "pi"; "log2"; "log10"; "sqrt2"; "not_a_number"; 
+    "positive_infinity"; "negative_infinity"; "machine_precision"; 
+    "abs"; "acos"; "acosh"; "asin"; "asinh"; "atan"; "atan2"; "atanh"; 
+    "cbrt"; "ceil"; "cos"; "cosh"; "erf"; "erfc"; "exp"; "exp2"; "expm1"; 
+    "fabs"; "floor"; "lgamma"; "log"; "log1p"; "log2"; "log10"; "round"; 
+    "sin"; "sinh"; "sqrt"; "tan"; "tanh"; "tgamma"; "trunc"; 
+    "fdim"; "fmax"; "fmin"; "hypot"; "fma" ]
+
+let contains_elt lst elt  = (List.mem lst elt ~equal:(=))
+
+let stan_namespace_qualify f = 
+  if (Set.mem functions_requiring_namespace f) then "stan::math::"^f else f
+
+(* return true if the types of the two expression are the same *)
+let types_match e1 e2 = 
+  e1.texpr_type = e2.texpr_type
+
+(* "__" is an illegal suffix for user functions, used for built-in operators not in signatures *)
+let is_user_defined f = 
+  not (Stan_math_signatures.is_stan_math_function_name f)
+  && (not (ends_with "__" f))
+  && (not (starts_with "stan::math::" f))
+
+(* retun true if the tpe of the expression is integer or real *)
+let is_scalar e = e.texpr_type = 
+  Ast.UInt || e.texpr_type = Ast.UReal
+
+let is_matrix e = 
+  e.texpr_type = Ast.UMatrix
+
+let is_row_vector e =
+  e.texpr_type = Ast.URowVector
+
+(* stub *)
+let pretty_print _e = "pretty printed e"
+
+(* stub *)
+let rec gen_return_type = function
+  | Ast.Void -> "void"
+  | Ast.ReturnType(rt) -> gen_type_ut rt
+
+(* terrible quadratic implementation *)
+and gen_arg_types = function
+  | [] -> ""
+  | [(_adt, ut)] -> (gen_type_ut ut)
+  | (_adt, ut)::ts -> sprintf "%s, %s" (gen_type_ut ut) (gen_arg_types ts)
```

Comment:
With Fmt this is much easier... Even without I think we can use some stdlib stuff:
```ocaml
let gen_arg_types ats = (String.concat ~sep:", " (List.map ~f:gen_type_ut ats))
or
let gen_arg_types ats = List.map ~f:gen_type_ut ats |> String.concat ~sep:", "
```